### DISCUSSION TITLE
How does xarray write a dask array to zarr using the distributed scheduler?

### DISCUSSION THREAD

**Discussion by ivirshup** [pydata/xarray#8115] [Q&A]
Hey,

This is much more of an implementation question than anything else.

Over on anndata we ran into the issue of serializing locks when trying to write a dask array to a hdf5 or zarr store:

* https://github.com/scverse/anndata/pull/1079

We came up with a solution based on what xarray does for HDF5, though in our case this only works for zarr I believe since we don't have xarrays `CachingFileManager` (https://github.com/pydata/xarray/issues/4242). However, xarray is not currently using this approach for `zarr` as far as I can tell.

**Question:** Could someone explain how xarray writes a dask array to a zarr store when using the distributed schedueler? I'm not seeing an obvious use of a lock or synchronizer in the zarr backend code.

**Comment by ivirshup** [comment]
Looking into it a bit, I believe xarray ends up passing `dask.store(..., lock=None)` for the zarr backend. Still working on figuring out where the chunks are aligned to make it safe to write without a lock though.

**Reply by TomNicholas** [reply]
The writing should all go through `.store` yes. For the other questions @jhamman might know.
